tools/generate-changelog.py

```diff
@@ -0,0 +1,167 @@
+#!/usr/bin/python3
+#
+# Copyright (C) 2019 Canonical Ltd
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+# Script that compares two manifest.yaml files and creates a paragraph
+# that includes the changes for all deb files. These changes are
+# obtained from the debian changelog for the different packages.
+
+import debian.changelog
+import debian.debian_support
+import gzip
+import requests
+import sys
+import yaml
+from collections import namedtuple
+
+
+def eprint(*args, **kwargs):
+    print(*args, file=sys.stderr, **kwargs)
+
+
+# Returns a dictionary from package name to version, using
+# the packages section.
+# manifest_p: path to manifest to load
+def packages_from_manifest(manifest_p):
+    with open(manifest_p) as manifest:
+        manifest_y = yaml.safe_load(manifest)
+
+        pkg_dict = {}
+        for pkg in manifest_y['packages']:
+            # package_name=version
+            pkg_data = pkg.split('=')
+            pkg_dict[pkg_data[0]] = pkg_data[1]
+        return pkg_dict
+
+def package_name(pkg):
+    t = pkg.split(':')
+    return t[0]
+
+def get_changelog_from_file(docs_d, pkg):
+    chl_path = docs_d + '/' + package_name(pkg) + '/changelog.Debian.gz'
+    with gzip.open(chl_path) as chl_fh:
+        return chl_fh.read().decode('utf-8')
+
+
+def get_changelog_from_url(pkg, new_v):
+    url = 'https://changelogs.ubuntu.com/changelogs/binary/'
+    safe_name = package_name(pkg)
+    if safe_name.startswith('lib'):
+        url += safe_name[0:4]
+    else:
+        url += safe_name[0]
+    url += '/' + safe_name + '/' + new_v + '/changelog'
+    changelog_r = requests.get(url)
+    if changelog_r.status_code != requests.codes.ok:
+        raise Exception('No changelog found in ' + url + ' - status:' +
+                        str(changelog_r.status_code))
+
+    return changelog_r.text
+
+
+# Gets difference in changelog between old and new versions
+# Returns source package and the differences
+def get_changes_for_version(docs_d, pkg, old_v, new_v, indent):
+    # Try to get changelog from file (only option that will work for
+    # ESM packages), otherwise go to changelogs.ubuntu.com.
+    try:
+        changelog = get_changelog_from_file(docs_d, pkg)
+    except Exception:
+        changelog = get_changelog_from_url(pkg, new_v)
```

Comment:
Do we really need that? I am not a fan of downloading anything outside of the "pull" phase in snapcraft.

Reply:
I do not know how some packages end of not having their changelog in the form of `changelog.Debian.gz` - but unfortunately it happens and the url function is hit - and correctly resolves the missing changelogs, so I'd prefer we keep it for now

Reply:
We know why it falls back here now - it's because we clean up in docs in the hooks :-)